sdk/appconfiguration/azure-appconfiguration-provider/azure/appconfigurationprovider/_azureappconfigurationprovider.py

```diff
@@ -0,0 +1,183 @@
+                key_filter=select.key_filter, label_filter=select.label_filter)
+            for config in configurations:
+                if config.content_type is None:
+                    # Deals with possible null value via Rest API
+                    provider._dict[provider.__trim(config.key)] = config.value
+                elif config.content_type == "application/vnd.microsoft.appconfig.keyvaultref+json;charset=utf-8":
+                    provider.__resolve_keyvault_references(
+                        config, key_vault_options, secret_clients)
+                elif "application/json" in config.content_type:
+                    j_object = json.loads(config.value)
+                    provider._dict[provider.__trim(config.key)] = j_object
+                else:
+                    provider._dict[provider.__trim(config.key)] = config.value
+        return provider
+
+    def __buildprovider(self, connection_string, endpoint, credential, key_vault_options):
+        headers = {}
+        correlation_context = "RequestType=Startup"
+
+        if (key_vault_options is not None and
+                (key_vault_options.credential is not None or key_vault_options.secret_clients is not None or
+                 key_vault_options.secret_resolver is not None)):
+            correlation_context += ",UsesKeyVault"
+
+        headers["Correlation-Context"] = correlation_context
+        useragent = USER_AGENT
+
+        if connection_string is not None:
+            self._client = AzureAppConfigurationClient.from_connection_string(
+                connection_string, user_agent=useragent, headers=headers)
+            return
+        self._client = AzureAppConfigurationClient(
+            endpoint, credential, user_agent=useragent, headers=headers)
+
+    def __resolve_keyvault_references(self, config, key_vault_options, secret_clients):
+        if key_vault_options is None:
+            raise AttributeError(
+                "Key Vault options must be set to resolve Key Vault references.")
+        j_object = json.loads(config.value)
+        uri_value = j_object['uri']
+
+        uri = urlparse(uri_value)
+
+        key_vault_uri = "https://" + uri.hostname
+        key_vault_secret_prefix = "/secrets/"
+        key_vault_secret_name = uri.path[len(
+            key_vault_secret_prefix):]
+        if key_vault_options.credential is not None:
+            secret_client = None
+
+            # Clients only should be made once, will reuse if client already made
+            for client_uri in secret_clients:
+                if client_uri == key_vault_uri:
+                    secret_client = secret_clients[client_uri]
+                    break
+            if secret_client is None:
+                secret_client = SecretClient(
+                    vault_url=key_vault_uri, credential=key_vault_options.credential)
+                secret_clients[key_vault_uri] = secret_client
+                secret = secret_client.get_secret(
```

Comment:
We need error handling around `get_secret`

Reply:
Should we combine these two gets so we don't double up on the logic. Trying to think of a good way to do this.

Reply:
Absolutely. This is the basic logic we have in .NET config provider:
We keep a list of secret clients - this includes user created clients and the ones we create using user provided credentials.
When we are trying to resolve a secret, we first search in our list of secret clients to see if there is a matching client already added. If yes, use it. If not, create a new one using user's credentials and add it to the list.

If there is no matching client found after the above steps, we check if there is any secret resolver. If yes, use secret resolver. If no, throw exception.

If there is a matching client, use that client to call get_secret.

Here's the code (you can ignore the secret cache logic because we dont have it here right now): https://github.com/Azure/AppConfiguration-DotnetProvider/blob/1f3f65bb6aa94f66ab62085d87f794e376585e0c/src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureKeyVaultReference/AzureKeyVaultSecretProvider.cs#L48-L70

Reply:
Interesting. I don't know why I did it this way. Spring works similar to what you gave above.

Reply:
@avanigupta can you verify this is the correct setup now?